---
title: "Introduction to Principal Component Region Set Analysis"
author: "John Lawson"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to Principal Component Region Set Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Purpose

Principal Component Region Set Analysis (PCRSA) is a method for annotating principal component scores using region sets to better interpret variation among samples. PCA reduces dimensions of the data and summarizes important features of data in a new set of dimensions called principal components (PCs). Currently, PCRSA is designed for understanding DNA methylation data (although it can potentially be used for other types of data like ATACseq). PCA is commonly used for DNA methylation data, but the PCs from DNA methylation data are hard to interpret because of the sheer number of CpG sites included and the ambiguity in function of each individual site. PCRSA annotates PCs with region sets, giving insight into variation in the data. A region set is a set of genomic regions that share a biological annotation. This includes transcription factor (TF) binding regions (eg from ChIP-seq), regions with a certain histone modification (eg ChIP-seq) or chromatin accessibility regions (eg DNase/ATAC-seq). Since DNA methylation can correlate with TF binding and with certain histone modifications, finding a region set to have variation in DNA methylation along the PC axis could also imply functional variation of the TF/histone modification/chromatin accessibility along the PC axis. By annotating PCs with region sets, you can identify meaningful sources of variation in your data, increasing your understanding of the epigenetic variation and the biological entity you are studying.  

# Our test case

```{r, eval=TRUE, message=FALSE, warning=FALSE, echo=FALSE}
library(PCRSA)
library(ggplot2)
data("brcaPCScores657")
ggplot(data = brcaPCScores657, mapping = aes(x=PC1, y=PC4, col=ER_Status)) + geom_point() + ggtitle("PCA of DNA Methylation Data from Breast Cancer Patients")
```

For this vignette, we are using data from The Cancer Genome Atlas: 450k DNA methylation microarrays for 657 breast cancer patients (TCGA-BRCA, https://portal.gdc.cancer.gov/). This vignette begins after PCA has been performed on the DNA methylation data. As you can see above, PC1 and PC4 cluster patients based on estrogen receptor status which is an important prognostic marker for breast cancer. Based on this, we might expect that these PCs would be associated with estrogen receptor. We are using an easy example for this vignette but PCRSA will still work even if there are not distinct clusters in your PCA plot. We will test a few region sets to see which ones have the highest association with our top PCs.

# Inputs

There are three main inputs:

1. PCA loadings (prcomp()$rotation).
2. Genomic coordinates associated with PCA loadings.
3. Region sets (normally a database of region sets).

PCA was done with DNA methylation data for all autosomal chromosomes but only the PC loadings and coordinates for cytosines in chromosome 1 are included in this vignette to minimize memory requirements. `brcaLoadings1` has chr1 loadings and corresponding coordinates are in `brcaCoord1`. In this vignette, we will use a small number of region sets to demonstrate how to use the PCRSA functions. For actual use, PCRSA should be done with a large number of region sets. Publicly available collections of region sets can be found online (eg http://databio.org/regiondb).

# Running PCRSA

First, we load the example data and required packages:

```{r, eval=TRUE, message=FALSE, warning=FALSE}
library(PCRSA)
library(data.table)
library(GenomicRanges)
library(ggplot2)
data("esr1_chr1")
data("gata3_chr1")
data("nrf1_chr1")
data("atf3_chr1")
data("brcaLoadings1")
data("brcaCoord1")
data("brcaPCScores")
data("brcaMethylData1")
```

We have region sets for four transcription factors: Esr1, Gata3, Nrf1, and Atf1.

```{r, eval=TRUE, message=FALSE, warning=FALSE}
# prepare data
GRList <- GRangesList(esr1_chr1, gata3_chr1, nrf1_chr1, atf3_chr1) 
rsName <- c("esr1_chr1", "gata3_chr1", "nrf1_chr1", "atf3_chr1")
```

Now we see how much each region set is associated with each principal component:

```{r, eval=TRUE, message=FALSE, warning=FALSE}
PCsToAnnotate <- paste0("PC", 1:4)
rsScores <- pcRegionSetEnrichment(loadingMat=brcaLoadings1, 
                                              mCoord=brcaCoord1, 
                                              GRList=GRList, 
                                              PCsToAnnotate=PCsToAnnotate, 
                                              scoringMetric="rsMean")
rsScores$rsName <- rsName
```

As an easy way to visualize the results, we can see how the region sets are ranked for each PC:

```{r, eval=TRUE, message=FALSE, warning=FALSE}
rsScoreHeatmap(rsScores, 
               PCsToAnnotate=paste0("PC", 1:4), 
               orderByPC = "PC1", column_title = "Region sets ordered by score for PC1")
```

We can see that Gata3 had the highest score for PCs 1, 3, and 4 but that Esr1 had a higher score for PC2. While you might expect that Esr1 would be ranked first for all PCs, these results actually make sense because Gata3 is known to be associated with breast cancer and Esr1 activity. On a practical note, if you want to arrange the heatmap by region set scores for another PC, you can just change the orderByPC parameter like so:

```{r, eval=TRUE, message=FALSE, warning=FALSE}
rsScoreHeatmap(rsScores, 
               PCsToAnnotate=paste0("PC", 1:4), 
               orderByPC = "PC2", column_title = "Region sets ordered by score for PC2")
```

# Understanding the Results

We can further understand the variability in these region sets in several ways:

1. Look at whether variability is specific to the regions of interest compared to the genome around these regions. 
2. Look at the raw DNA methylation in these regions and whether it follows the same trends as the PC scores.
3. Look at the loadings in each region of a given region set to see whether all regions have high loadings or just a subset the regions. Also we can see if the same regions have high loadings for multiple PCs.

## Specificity of variation to the regions of interest

We can see whether variability along the PC is specific to the region of interest by comparing the region of interest to the surrounding genome. To do this, we will calculate the average loadings of a wide area surrounding the regions of interest.

```{r, eval=TRUE, message=FALSE, warning=FALSE}
wideGRList <- lapply(GRList, resize, width=14000, fix="center")
loadProfile <- pcEnrichmentProfile(loadingMat=brcaLoadings1, mCoord=brcaCoord1,
                                 GRList=wideGRList, PCsToAnnotate=PCsToAnnotate,
                                 binNum=21)
```

We will normalize the result for each PC so we can better compare them. Here we normalize by subtracting the mean absolute loading of each PC from the region set profiles for the corresponding PC. Then we get the plot scale so we can easily compare the different profiles. 

```{r, eval=TRUE, message=FALSE, warning=FALSE}
# average loading value from each PC to normalize so PCs can be compared with each other
avLoad <- apply(X=brcaLoadings1[, PCsToAnnotate], MARGIN=2, FUN=function(x) mean(abs(x)))

# normalize
loadProfile <- lapply(loadProfile, FUN=function(x) x[, mapply(FUN=function(y, z) get(y) - z, y=PCsToAnnotate, z=avLoad)])
loadProfile <- lapply(loadProfile, FUN=function(x) data.table(regionGroupID=1:nrow(x), x))

# for the plot scale
maxVal <- max(sapply(loadProfile, FUN=function(x) max(x[, .SD, .SDcols=PCsToAnnotate])))
minVal <- min(sapply(loadProfile, FUN=function(x) min(x[, .SD, .SDcols=PCsToAnnotate])))

# convert to long format for plots
loadProfile <- lapply(X=loadProfile, FUN=function(x) tidyr::gather(data=x, key="PC", value="loading_value", PCsToAnnotate))
loadProfile <- lapply(X=loadProfile, as.data.table)
loadProfile <- lapply(loadProfile, function(x) x[, PC := factor(PC, levels=PCsToAnnotate)])
```

Let's look at the plots!

```{r, eval=TRUE, message=FALSE, warning=FALSE}
wrapper <- function(x, ...) paste(strwrap(x, ...), collapse="\n") 
profilePList <- list()
for (i in seq_along(loadProfile)) {
    
    thisRS <- loadProfile[[i]]
    
    profilePList[[i]] <- ggplot(data=thisRS, mapping=aes(x=regionGroupID , y=loading_value)) + 
        geom_line() + ylim(c(minVal, maxVal)) + facet_wrap(facets="PC") + 
        ggtitle(label=wrapper(rsName[i], width=30)) + xlab("Genome around Region Set, 14 kb") + 
        ylab("Normalized Loading Value") + 
        theme(panel.grid.major.x=element_blank(), panel.grid.minor.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank())
    profilePList[[i]]

}
profilePList[[1]]
profilePList[[2]]
profilePList[[3]]
profilePList[[4]]
```

These plots show that Esr1 and Gata3 binding regions are variable compared to the surrounding genome (because they have a peak in the middle) while Nrf1 and Atf3 are not. Some region sets may have an increased loading but no peak, for example some histone modification region sets like H3K27me3 or H3K9me3. This doesn't necessarily mean these regions are not relevant. It could just mean that there are larger blocks of variability around these histone modifications. The results from our four region sets suggest that Esr1 and Gata3 regions contribute to the variation along the PCs 1 and 3 (as well as perhaps 4), helping us understand the biological meaning of those PCs (at least partially estrogen receptor-related) and understand sources of variation in our data.

## Looking at the Raw Data

If a region set has a high score for a PC, we would expect that DNA methylation in at least some of those regions would correlate with PC score. In other words, as you go from high PC score to low PC score along the PC axis, DNA methylation will either go up or down. Let's look at DNA methylation at CpGs in Esr1 regions. In the following plot, each column is a CpG in an Esr1 region and each row is a patient. Patients are ordered by their PC score for PC1.

```{r, eval=TRUE, message=FALSE, warning=FALSE}
featuresAlongPC(methylData=brcaMethylData1,
                mCoord=brcaCoord1,
                regionSet=esr1_chr1,
                pcScores=brcaPCScores,
                orderByPC="PC1", cluster_columns=TRUE, name = "DNA Methylation Level")
```

It appears that some but not all CpGs vary greatly along the PC axis. The CpGs that show high variation along the PC axis are the ones that contribute to the Esr1 region set being ranked highly in our analysis. Looking at the raw data confirms that DNA methylation is in fact varying along the PC axis, which gives us more confidence in our results.

Now let's look at one of the region sets, Nrf1, that had a low score for PC1:

```{r, eval=TRUE, message=FALSE, warning=FALSE}
featuresAlongPC(methylData=brcaMethylData1,
                mCoord=brcaCoord1,
                regionSet=nrf1_chr1,
                pcScores=brcaPCScores,
                orderByPC="PC1", cluster_columns=TRUE, name = "DNA Methylation Level")
```

We can see that there is very little variation in DNA methylation, if any, along the PC1 axis. Therefore, it is not surprising that the Nrf1 region set had a low score for this PC.

Since PCRSA ranks region sets based on their relative scores in comparison to other region sets tested, there will always be a region set with a best score. Therefore, it is always a good idea to check the raw DNA methylation in your top region sets to make sure that there really is variation along the PC.

## Looking at Loadings of Individual Regions

This plot can help you learn more about the contribution of individual regions to the region set score for each PC. For example, if the estrogen receptor region set was associated with PCs 1, 3, and 4, we might wonder whether the same regions are causing the association with these PCs or whether different regions are associated with each PC. To do this, we will first calculate the average absolute loading value for each region in a region set (obtained by averaging a given PC's loadings for CpGs within that region). Then we can use the distribution of loadings for each PC to convert each region's loading to a percentile to see how extreme/high that region is for each PC. Let's look at the plot for estrogen receptor: 

```{r, eval=TRUE, message=FALSE, warning=FALSE}
regionQuantileByPC(loadingMat = brcaLoadings1,
                                mCoord = brcaCoord1,
                                regionSet = esr1_chr1,
                                rsName = "Estrogen Receptor (Chr1)",
                                PCsToAnnotate=paste0("PC", 1:4),
                                maxRegionsToPlot = 8000,
                                cluster_rows = TRUE,
                                cluster_columns = FALSE,
                                column_title = rsName,
                                name = "Percentile of Loading Scores in PC")
```

We can see that some of the same regions have high loadings for multiple PCs (ie these regions are important for these PCs). Also, there are some regions that do not have high loadings for any of the top 4 PCs, suggesting that these regions are not associated with the largest sources of covariation in the data.  Overall, PC2 does not have as high loadings as PCs 1, 3, and 4, consistent with our loading profiles (no peak for PC2). While Esr1 was ranked highest for PC2 for the initial PCRSA results, this was only the highest score out of 4 region sets. This illustrates two important points for using PCRSA. First, you almost always will want to use many region sets so that you can compare the PCRSA score for a region set to many others to see if it really is relatively high. Second, you should do further confirmation on top-ranked region sets such as looking at the raw DNA methylation values and looking at the loading profiles to make sure your top region sets are capturing real variation.  

For contrast, we can look at the regions of Nrf1:

```{r, eval=TRUE, message=FALSE, warning=FALSE}
regionQuantileByPC(loadingMat = brcaLoadings1,
                                mCoord = brcaCoord1,
                                regionSet = nrf1_chr1,
                                rsName = "Nrf1 (Chr1)",
                                PCsToAnnotate=paste0("PC", 1:4),
                                maxRegionsToPlot = 8000,
                                cluster_rows = TRUE,
                                cluster_columns = FALSE,
                                column_title = rsName,
                                name = "Percentile of Loading Scores in PC")
```


We can see that only a very small proportion of regions in the Nrf1 region set have high loadings for PCs 1-4. This is consistent with this region set being ranked low by PCRSA for association with these PCs.

These visualization functions can help you explore your data and understand its variation. In conclusion, PCRSA can help you identify biologically meaningful sources of variation that contribute to each PC, giving you insight into variability in your data. 


# Related references
This paper involves gene sets instead of region sets but provided some inspiration for our method:

Frost HR, Li Z, Moore JH. Principal component gene set enrichment (PCGSE). BioData Mining. 2015;8:25. doi:10.1186/s13040-015-0059-z.
