---
title: "Supervised COCOA"
author: "John Lawson"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Supervised COCOA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Supervised COCOA

Goal: Understand epigenetic variation related to a specific phenotype

In a supervised COCOA analysis, we want to annotate epigenetic variation that is related to variation in one or more sample phenotypes. These sample phenotypes could be molecular (e.g. expression of a protein marker, presence/variant allele frequency of a mutation, or expression of a certain gene). The sample phenotypes can also be higher-level organism phenotypes (e.g. patient survival, cancer stage, disease severity). COCOA will identify region sets where the epigenetic signal is most related to the phenotype. These region sets then can help to understand the relationship between the phenotype and epigenetic state. 

Vignette analysis goal: Understand epigenetic variation related to our phenotype of interest, estrogen receptor (ER) status

# Quantify relationship between chosen sample phenotype and epigenetic data

First we will load the necessary data: region sets to test, BRCA DNA methylation data with genomic coordinates, and our phenotype of interest, ER status (part of `brcaPCScores657`)

```{r, eval=TRUE, message=FALSE, warning=FALSE}
library(COCOA)
data("esr1_chr1")
data("gata3_chr1")
data("nrf1_chr1")
data("atf3_chr1")
data("brcaMCoord1")
data("brcaMethylData1")
data("brcaPCScores657")
sampleLabels <- brcaPCScores657[colnames(brcaMethylData1), ]
```

Let's convert estrogen receptor status to a number so we can do calculations with it. Then we will quantify the association between ER status and the DNA methylation level at each cytosine using Pearson correlation.

```{r, eval=TRUE, message=FALSE, warning=FALSE}
sampleLabels$ER_Status <- scale(as.numeric(sampleLabels$ER_Status), 
                               center=TRUE, scale=FALSE)
methylCor <- cor(t(brcaMethylData1), sampleLabels$ER_Status)
# if the standard deviation of the methylation level 
# for a CpG across samples is 0,
# cor() will return NA, so manually set the correlation to 0 for these CpGs
methylCor[is.na(methylCor)] <- 0
colnames(methylCor) <- "ER_Status"
```

Now we have a feature contribution score (the correlation coefficient in this case) for each methylation site that represents how much the methylation at that site is associated with ER status. We can use those individual cytosine scores to score region sets. We'll score our region sets with the `runCOCOA` function.

# Score region sets

```{r, eval=TRUE, message=FALSE, warning=FALSE}
GRList <- GRangesList(esr1_chr1, gata3_chr1, atf3_chr1, nrf1_chr1)
rsNames <- c("esr1", "gata3", "atf3", "nrf1")
rsScores <- runCOCOA(signal=methylCor,
                     signalCoord=brcaMCoord1,
                     GRList=GRList,
                     signalCol="ER_Status",
                     scoringMetric="regionMean",
                     absVal=TRUE)
rsScores
```

Now we have a score for each region set that represents how much epigenetic variation in that region set is associated with the phenotype of interest, ER status.

Both the correlation and scoring steps can be combined into one step with the `corPerm` function as will be shown in the next section. 

# Estimating statistical significance

Now that we have a score for each region set, we want to get an idea of the statistical significance of

## Permutations

We need to set the seed for reproducible results since we are doing random shuffling of the sample phenotypes.

```{r, eval=TRUE, message=FALSE, warning=FALSE}
set.seed(100)

nPerm <- 5
permRSScores <- list()

for (i in 1:nPerm) {
    # shuffling sample labels
    randomInd <- sample(1:nrow(sampleLabels), nrow(sampleLabels))
    permRSScores[[i]] <- corPerm(randomInd=randomInd, 
                           genomicSignal=brcaMethylData1, 
                           signalCoord=brcaMCoord1, 
                           GRList=GRList,
                           calcCols="ER_Status", 
                           sampleLabels=sampleLabels, 
                           variationMetric="cor")
}

permRSScores

```

We now have a list where each item is a result of runCOCOA but we'll reformat it so that we have a list where each item is a null distribution for a single region set. We have `nPerm` list items in permRSScores.

```{r, eval=TRUE, message=FALSE, warning=FALSE}
nullDistList <- convertToFromNullDist(permRSScores)
nullDistList
```

We have four null distributions, one for each region set in `GRList`.

## Fit gamma distribution to null distributions and get p-values

To reduce the number of permutations (and the run time for COCOA), we will approximate each null distribution with a gamma distribution [citation]. After fitting the gamma distribution to each null distribution, we can use the gamma distributions to get p-values. 

```{r, eval=TRUE, message=FALSE, warning=FALSE}
# p-values based on fitted gamma distributions
        gPValDF <- getGammaPVal(rsScores <- rsScores, 
                               nullDistList = nullDistList, 
                               calcCols = "ER_Status", 
                               method = "mme", realScoreInDist = TRUE)
        gPValDF <- cbind(gPValDF, 
                        rsScores[, colnames(rsScores)[!(colnames(rsScores) 
                                                                %in% "ER_Status")]])
        gPValDF <- cbind(gPValDF, rsNames)
        gPValDF
```


# How to choose a metric?
    
